### ISSUE TITLE
Regression: 1.2.0 and 1.2.1 fail to build from source, while 1.2.0-beta1 is OK

### ISSUE THREAD

**Issue by sten0** [hydrogen-music/hydrogen#1798]
Hydrogen version *          : 1.2.0 and 1.2.1
Operating system + version  :  Debian sid/unstable or Debian 12 (bookworm)
Audio driver + version      :  NA

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Thank you for your continued work on Hydrogen.  I attempted to build 1.2.0 (final release) the 23rd of May, as well as 1.2.1 the 13th of June.  The objective is of course to update Debian's package, enable drumkit upgrades for system-managed kits, etc.  The following is the most relevant snippet from the log of a failed build of 1.2.1.  1.2.0-beta1 is good, and builds of 1.2.0 when no other variables are introduced.
```
[ 34%] Building CXX object src/core/CMakeFiles/hydrogen-core-1.2.1.dir/Lilipond/Lilypond.cpp.o
cd /<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core && /usr/bin/c++ -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_XMLPATTERNS_LIB
-DQT_XML_LIB -I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core -I/<<PKGBUILDDIR>>/src/core
-I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core/hydrogen-core-1.2.1_autogen/include -I/<<PKGBUILDDIR>>/src/core/include
-I/<<PKGBUILDDIR>>/src -I/<<PKGBUILDDIR>>/include -I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core/.. -I/usr/include/opus
-I/usr/include/raptor2 -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -isystem
/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem
/usr/include/x86_64-linux-gnu/qt5/QtXmlPatterns -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem
/usr/include/x86_64-linux-gnu/qt5/QtGui -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat
-Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O0 -fno-implement-inlines -g  -std=gnu++17 -fPIC -MD -MT
src/core/CMakeFiles/hydrogen-core-1.2.1.dir/Lilipond/Lilypond.cpp.o -MF CMakeFiles/hydrogen-core-1.2.1.dir/Lilipond/Lilypond.cpp.o.d
-o CMakeFiles/hydrogen-core-1.2.1.dir/Lilipond/Lilypond.cpp.o -c /<<PKGBUILDDIR>>/src/core/Lilipond/Lilypond.cpp
In file included from /<<PKGBUILDDIR>>/src/core/IO/MidiInput.h:26,
                 from /<<PKGBUILDDIR>>/src/core/IO/PortMidiDriver.h:26,
                 from /<<PKGBUILDDIR>>/src/core/IO/PortMidiDriver.cpp:24:
/<<PKGBUILDDIR>>/src/core/IO/PortMidiDriver.cpp: In member function ‘virtual void H2Core::PortMidiDriver::open()’:
/<<PKGBUILDDIR>>/src/core/IO/PortMidiDriver.cpp:220:55: error: ‘const struct PmDeviceInfo’ has no member named ‘is_virtual’
  220 |                                          .arg( pInfo->is_virtual == TRUE ? "virtual " : "" )
      |                                                       ^~~~~~~~~~
/<<PKGBUILDDIR>>/src/core/Object.h:229:163: note: in definition of macro ‘__LOG_METHOD’
  229 | #define __LOG_METHOD(   lvl, msg )  if( __logger->should_log( (lvl) ) )                 { __logger->log( (lvl), _class_name(),
__FUNCTION__, QString( "%1" ).arg( msg ) ); }

|                                                                                                                                      \
                             ^~~
/<<PKGBUILDDIR>>/src/core/IO/PortMidiDriver.cpp:217:25: note: in expansion of macro ‘INFOLOG’
  217 |                         INFOLOG( QString( "%1%2%3%4device called [%5] using [%6] MIDI API" )
      |                         ^~~~~~~
[ 35%] Building CXX object src/core/CMakeFiles/hydrogen-core-1.2.1.dir/Logger.cpp.o
cd /<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core && /usr/bin/c++ -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_XMLPATTERNS_LIB
-DQT_XML_LIB -I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core -I/<<PKGBUILDDIR>>/src/core
-I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core/hydrogen-core-1.2.1_autogen/include -I/<<PKGBUILDDIR>>/src/core/include
-I/<<PKGBUILDDIR>>/src -I/<<PKGBUILDDIR>>/include -I/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu/src/core/.. -I/usr/include/opus
-I/usr/include/raptor2 -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -isystem
/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem
/usr/include/x86_64-linux-gnu/qt5/QtXmlPatterns -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem
/usr/include/x86_64-linux-gnu/qt5/QtGui -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat
-Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O0 -fno-implement-inlines -g  -std=gnu++17 -fPIC -MD -MT
src/core/CMakeFiles/hydrogen-core-1.2.1.dir/Logger.cpp.o -MF CMakeFiles/hydrogen-core-1.2.1.dir/Logger.cpp.o.d -o
CMakeFiles/hydrogen-core-1.2.1.dir/Logger.cpp.o -c /<<PKGBUILDDIR>>/src/core/Logger.cpp
make[4]: *** [src/core/CMakeFiles/hydrogen-core-1.2.1.dir/build.make:681:
src/core/CMakeFiles/hydrogen-core-1.2.1.dir/IO/PortMidiDriver.cpp.o] Error 1
make[4]: *** Waiting for unfinished jobs....
make[4]: Leaving directory '/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu'
make[3]: *** [CMakeFiles/Makefile2:250: src/core/CMakeFiles/hydrogen-core-1.2.1.dir/all] Error 2
make[3]: Leaving directory '/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu'
make[2]: *** [Makefile:159: all] Error 2
make[2]: Leaving directory '/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu'
dh_auto_build: error: cd obj-x86_64-linux-gnu && make -j4 "INSTALL=install --strip-program=true" VERBOSE=1 returned exit code 2
make[1]: *** [debian/rules:20: override_dh_auto_build] Error 25
make[1]: Leaving directory '/<<PKGBUILDDIR>>'
make: *** [debian/rules:13: binary] Error 2
dpkg-buildpackage: error: debian/rules binary subprocess returned exit status 2
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

*  H2 version : Release tarballs for 1.2.0-beta1, 1.2.0, and 1.2.1
** To reproduce, build on any version of Debian with GCC 12.3.0-4.  It may be necessary to add "-fstack-protector-strong -Wformat
-Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2" to CMAKE_CXX_FLAGS and/or CMAKE_EXE_LINKER_FLAGS.

Here is the full log, and all relevant packages are listed during the installation of dependencies phase.  I'm also attaching the log from 1.2.0 which has a different error that appears to have been fixed in 1.2.1.  Log of 1.2.0-beta1 (on the same build system) is available on request.
[hydrogen_1.2.1+dfsg-1~exp1_amd64.build.txt](https://github.com/hydrogen-music/hydrogen/files/11783691/hydrogen_1.2.1%2Bdfsg-1.exp1_amd64.build.txt)
[hydrogen_1.2.0+dfsg-1_amd64.build.txt](https://github.com/hydrogen-music/hydrogen/files/11783702/hydrogen_1.2.0%2Bdfsg-1_amd64.build.txt)

Recent logs of 1.2.0-beta1 one on Debian infrastructure are available here:
https://ci.debian.net/packages/h/hydrogen/unstable/amd64/
https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/hydrogen.html

'hope this is an easy fix rather than an involved one!
Nicholas

**Comment by theGreatWhiteShark** [comment]
Hey @sten0 ,

Yes, this one was already reported (#1795). Turns out a line containing features of a more modern version of PortMidi slipped into the code.

There are two things we can do here. Either you compile it without PortMidi (its just a wrapper around ALSA without supporting all its features (yet) and we mainly use it to support MIDI on Windows) or I patch the code in `releases/1.2` branch to compile with PortMidi 217 again.

The patch would be very tiny as it just affects a log message. However, version 1.3. of Hydrogen will require a version of PortMidi >= v2.0.1. There is a lot of movement in this project lately and it will allow to align the UX for both ALSA and PortMidi driver within Hydrogen. But v1.3. of Hydrogen won't most probably not happen before 2024. 

P.S.
The 1.2.0 bug should also be fixed by now (see #1746)

**Comment by sten0** [comment]
Hi @theGreatWhiteShark ,

theGreatWhiteShark ***@***.***> writes:

> Hey @sten0 ,
>
> Yes, this one was already reported (#1795). Turns out a line containing features of a more modern version of PortMidi slipped into the code.

Sorry, I confess that I didn't look hard enough, and thank you for the
explanation and link.

> There are two things we can do here. Either you compile it without PortMidi (its just a wrapper around ALSA without supporting all its features (yet) and we mainly use it to support MIDI on Windows) or I patch the code in `releases/1.2` branch to compile with PortMidi 217 again.
>
> The patch would be very tiny as it just affects a log message.

Yes, please, I would appreciate this tiny patch, and I'm relieved that
it's a small amount of work :) The reason I believe that this approach
is best is because it minimises disruption all of the LTS users.  Please
let me know if you'd like the extended rationale (I'm guessing you'd
rather not spend time to read it).

> However, version 1.3. of Hydrogen will require a version of PortMidi >= v2.0.1. There is a lot of movement in this project lately and it will allow to align the UX for both ALSA and PortMidi driver within Hydrogen. But v1.3. of Hydrogen won't most probably not happen before 2024. 

I just filed a request for PortMIDI >= v2.0.4 in Debian.  If a prelease
of Hydrogen 1.3 isn't in Debian in 2024, please ping me, and I'll
probably be able to take care of this dependency too.

> P.S.
> The 1.2.0 bug should also be fixed by now (see #1746)

Thank you for the confirmation, much appreciated!

**Comment by theGreatWhiteShark** [comment]
> Yes, please, I would appreciate this tiny patch, and I'm relieved that
it's a small amount of work

Should be fixed on branch `releases/1.2` now.

> I'm guessing you'd rather not spend time to read it

That sounds way too dire and I certainly will find time to read what you post here.

>  If a prelease
of Hydrogen 1.3 isn't in Debian in 2024, please ping me, and I'll
probably be able to take care of this dependency too.

I don't have any time schedule. There is a rather big new feature (#1691)  coming which will need at least one round of beta testing and several smaller fixes pending. I will work on it whenever I find some time and it's done when it's done. :)

**Comment by theGreatWhiteShark** [comment]
should be fixed in release 1.2.2
